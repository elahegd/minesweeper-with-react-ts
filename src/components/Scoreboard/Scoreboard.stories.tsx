import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Scoreboard, ScoreboardProps } from './Scoreboard';

export default {
    title: "Scoreboard/Scoreboard",
    component: Scoreboard
} as Meta;

const Template: Story<ScoreboardProps> = (args) => <Scoreboard {...args} />;

export const ScoreboardExample = Template.bind({});
ScoreboardExample.args = {
    timer: "000",
    levels: ['beginner', 'intermediat', 'expert'],
    defaultValue: 'beginner',
    bombs: "010",
}