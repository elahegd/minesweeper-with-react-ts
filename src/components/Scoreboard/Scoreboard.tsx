import React, { FC, ChangeEvent } from 'react';
import styled from "@emotion/styled";

import { Reset } from "./Reset";
import { Counter } from "./Counter";
import { Level } from './Level';

export interface ScoreboardProps {
    timer: string;
    levels: string[];
    defaultValue?: string;
    onChangeLevel: (event: ChangeEvent<HTMLSelectElement>) => void;
    onReset: () => void;
    bombs: string;
}

export const Scoreboard: FC<ScoreboardProps> = ({timer, levels, defaultValue, onChangeLevel, onReset, bombs}) => {
  return (
    <Wrapper>
      <Counter>{timer}</Counter>
      <ResetAndLevelWrapper>
        <Level value={defaultValue} onChange={onChangeLevel}>
          {levels}
        </Level>
        <Reset onReset={onReset} />
      </ResetAndLevelWrapper>
      <Counter>{bombs}</Counter>
    </Wrapper>
  )
}

const Wrapper = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;
const ResetAndLevelWrapper = styled.div`
  display: flex;
`;