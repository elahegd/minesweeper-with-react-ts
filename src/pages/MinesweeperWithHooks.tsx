import React, { FC } from 'react';

import { useParams } from "react-router-dom";

import { GameLayout } from '../components/Game';
import  Top  from "../components/Top";
import { GameWithHooks } from '../module/GameWithHooks';

const MinesweeperWithHooks: FC = () => {
    const { username } = useParams<{ username?: string }>();

    return (
        <GameLayout 
            top={
                <Top feature="Flag" firstAction="ctrl" secondAction="click">
                    Minesweeper with React Hooks special for you: {username && ` ${username}`}
                </Top>
            }
        >
            <GameWithHooks />
        </GameLayout>
    );
}

export default MinesweeperWithHooks;