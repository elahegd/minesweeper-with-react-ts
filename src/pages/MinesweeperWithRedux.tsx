import React, { FC } from 'react';

import { GameLayout } from '../components/Game';
import  Top  from "../components/Top";
import { GameWithRedux } from '../module/GameWithRedux';

const MinesweeperWithRedux: FC = () => {
    return (
        <GameLayout 
            top={
                <Top feature="Flag" firstAction="ctrl" secondAction="click">
                    Minesweeper with React and Redux uses useReducer
                </Top>
            }
        >
            <GameWithRedux />
        </GameLayout>
    );
}

export default MinesweeperWithRedux;