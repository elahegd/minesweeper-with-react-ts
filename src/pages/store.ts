import { configureStore } from "@reduxjs/toolkit";
import { reducer } from "../module/GameWithRedux";

export const store = configureStore({
    reducer: {
        game: reducer
    }
});

export type RootState = ReturnType<typeof store.getState>;