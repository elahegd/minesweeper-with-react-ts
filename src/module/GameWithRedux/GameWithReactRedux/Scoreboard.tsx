import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RootState } from "../../../pages/store";
import { actions } from "../index";

import { Scoreboard as ScoreboardComponent } from "../../../components/Scoreboard";
import { GameLevel, LevelNames } from '../../GameSettings';

export const Scoreboard: FC = () => {
    const dispatch = useDispatch();

    const { time, bombs, flagCounter, level } = useSelector(
        ({ game: { time, bombs, flagCounter, level } }: RootState) => ({ 
            time, bombs, flagCounter, level 
        })
    );

    const handleChange = useCallback(
        ({target: {value: level}}: React.ChangeEvent<HTMLSelectElement>) => {
          dispatch(actions.changeLevel(level as LevelNames))
          }, 
          // Stryker disable next-line ArrayDeclaration
          []);

    const onReset = useCallback(
        () => dispatch(actions.reset()), 
        // Stryker disable next-line ArrayDeclaration
        []
    );

    return (
        <>
            <ScoreboardComponent 
                timer={String(time)} 
                levels={GameLevel as unknown as string[]}
                defaultValue={level}
                onChangeLevel={handleChange}
                onReset={onReset}
                bombs={String(bombs - flagCounter)}
            />
        </>
    )
}
