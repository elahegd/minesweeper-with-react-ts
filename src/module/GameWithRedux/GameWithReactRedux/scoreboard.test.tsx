import React from 'react';
import { Provider } from 'react-redux';
import { render } from "@testing-library/react";

import { store } from '../../../pages/store';

import { Scoreboard } from './Scoreboard';

describe("Scoreboard test cases", () => {
    it("Scoreboard renders correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Scoreboard />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
})