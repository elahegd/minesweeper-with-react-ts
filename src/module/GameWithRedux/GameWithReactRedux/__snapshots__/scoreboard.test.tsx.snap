// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard test cases Scoreboard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1n7y7g1"
  >
    <div
      class="css-d2q1pa"
    >
      0
    </div>
    <div
      class="css-ov1ktg"
    >
      <select
        class="css-1lpmfg0"
      >
        <option
          class="css-zcmdxr"
          value="beginner"
        >
          beginner
        </option>
        <option
          class="css-zcmdxr"
          value="intermediate"
        >
          intermediate
        </option>
        <option
          class="css-zcmdxr"
          value="expert"
        >
          expert
        </option>
      </select>
      <div>
        <button
          class="css-1kb3z2h"
        >
          ðŸ˜€
        </button>
      </div>
    </div>
    <div
      class="css-d2q1pa"
    >
      10
    </div>
  </div>
</DocumentFragment>
`;
